core/src/main/scala/kafka/utils/CoreUtils.scala

```diff
@@ -252,16 +255,60 @@ object CoreUtils {
     listenerListToEndPoints(listeners, securityProtocolMap, true)
   }
 
+  def validateOneIsIpv4AndOtherIpv6(first: String, second: String): Boolean =
+    (inetAddressValidator.isValidInet4Address(first) && inetAddressValidator.isValidInet6Address(second)) ||
+      (inetAddressValidator.isValidInet6Address(first) && inetAddressValidator.isValidInet4Address(second))
+
+  def checkDuplicateListenerPorts(endpoints: Seq[EndPoint], listeners: String): Unit = {
+    val distinctPorts = endpoints.map(_.port).distinct
+    require(distinctPorts.size == endpoints.map(_.port).size, s"Each listener must have a different port, listeners: $listeners")
+  }
+
   def listenerListToEndPoints(listeners: String, securityProtocolMap: Map[ListenerName, SecurityProtocol], requireDistinctPorts: Boolean): Seq[EndPoint] = {
     def validate(endPoints: Seq[EndPoint]): Unit = {
-      // filter port 0 for unit tests
-      val portsExcludingZero = endPoints.map(_.port).filter(_ != 0)
       val distinctListenerNames = endPoints.map(_.listenerName).distinct
-
       require(distinctListenerNames.size == endPoints.size, s"Each listener must have a different name, listeners: $listeners")
-      if (requireDistinctPorts) {
-        val distinctPorts = portsExcludingZero.distinct
-        require(distinctPorts.size == portsExcludingZero.size, s"Each listener must have a different port, listeners: $listeners")
+
+      val (duplicatePorts, _) = endPoints.filter {
+        // filter port 0 for unit tests
+        ep => ep.port != 0
+      }.groupBy(_.port).partition {
+        case (_, endpoints) => endpoints.size > 1
+      }
+
+      // Exception case, let's allow duplicate ports if one host is on IPv4 and the other one is on IPv6
+      val duplicatePortsPartitionedByValidIps = duplicatePorts.map {
+        case (port, eps) =>
+          (port, eps.partition(ep =>
+            ep.host != null && inetAddressValidator.isValid(ep.host)
+          ))
+      }
+
+      // Iterate through every grouping of duplicates by port to see if they are valid
+      duplicatePortsPartitionedByValidIps.foreach {
+        case (port, (duplicatesWithIpHosts, duplicatesWithoutIpHosts)) =>
+          if (requireDistinctPorts)
+            checkDuplicateListenerPorts(duplicatesWithoutIpHosts, listeners)
```

Comment:
I think we should throw exception if `duplicatesWithoutIpHosts` is not empty here, because we actually only expect `duplicatesWithIpHosts/duplicatesWithoutIpHosts` to be: `"2 IPs"/empty list`, right? After all, we only allow duplicate ports under **valid** IPv4 and IPv6, so we should be able to throw exception directly here, and remove the check below:
```
if (duplicatesWithoutIpHosts.nonEmpty)
      throw new IllegalArgumentException(errorMessage)
```
WDYT?

Something like this:
```
case (port, (duplicatesWithIpHosts, duplicatesWithoutIpHosts)) =>
  // It's invalid if listener on this port without an IP host (such as a null host) 
  if (duplicatesWithoutIpHosts.nonEmpty)
      throw new IllegalArgumentException("...")

  duplicatesWithIpHosts match {
    case eps ....
    if (requireDistinctPorts)
                require(validateOneIsIpv4AndOtherIpv6(ep1.host, ep2.host), "If you have two listeners on " +
                  s"the same port then one needs to be IPv4 and the other IPv6, listeners: $listeners, port: $port")
```

Reply:
Doing so causes this test to fail

```
props.put(KafkaConfig.ListenersProp, "PLAINTEXT://apache.org:9092,SSL://[::1]:9092")
caught = assertThrows(classOf[IllegalArgumentException], () => KafkaConfig.fromProps(props))
assertTrue(caught.getMessage.contains("Each listener must have a different port"))
```

Reply:
I checked, the `apache.org` will be seen as `duplicatesWithoutIpHosts`, and it'll enter here:
```
if (duplicatesWithoutIpHosts.nonEmpty)
      throw new IllegalArgumentException("...")
```
which should also throw exception, and pass the test, unless the error message is not expected. But, if you think my above suggestion doesn't look better, you can keep current implementation. Just want to let you know I think there's a better and more clear way to implement the logic. Thanks.

Reply:
> But, if you think my above suggestion doesn't look better, you can keep current implementation. Just want to let you know I think there's a better and more clear way to implement the logic. Thanks.

Don't disagree with the changes, just that they appear to be making a regression (i.e. breaking an existing test). Just to make sure something really weird isn't going on I attached a `.patch` file located within a zip so we are talking about the same change.

[reorder-exception.patch.zip](https://github.com/apache/kafka/files/9187405/reorder-exception.patch.zip)

Reply:
Agree we can address this comment in another PR.

I've checked the patch file, and found the following lines are not removed:
```
           if (requireDistinctPorts)
             checkDuplicateListenerPorts(duplicatesWithoutIpHosts, listeners)
```
Because we don't expect `duplicatesWithoutIpHosts` is nonEmpty, we don't need to validate them.

So, it will look like this:
```
case (port, (duplicatesWithIpHosts, duplicatesWithoutIpHosts)) =>
  val errorMessage = "If you have two listeners on " +
     s"the same port then one needs to be IPv4 and the other IPv6, listeners: $listeners, port: $port"
     if (duplicatesWithoutIpHosts.nonEmpty)
        throw new IllegalArgumentException(errorMessage)

      duplicatesWithIpHosts match {
            case eps if eps.isEmpty =>
            case Seq(ep1, ep2) =>
              if (requireDistinctPorts)
                require(validateOneIsIpv4AndOtherIpv6(ep1.host, ep2.host), errorMessage)
             case _ =>
               ....
```

Reply:
So this is exactly the version that I have

```scala
// Iterate through every grouping of duplicates by port to see if they are valid
duplicatePortsPartitionedByValidIps.foreach {
  case (port, (duplicatesWithIpHosts, duplicatesWithoutIpHosts)) =>
    val errorMessage = "If you have two listeners on " +
      s"the same port then one needs to be IPv4 and the other IPv6, listeners: $listeners, port: $port"
    if (duplicatesWithoutIpHosts.nonEmpty)
      throw new IllegalArgumentException(errorMessage)

    duplicatesWithIpHosts match {
      case eps if eps.isEmpty =>
      case Seq(ep1, ep2) =>
        if (requireDistinctPorts)
          require(validateOneIsIpv4AndOtherIpv6(ep1.host, ep2.host), errorMessage)
      case _ =>
        // Having more than 2 duplicate endpoints doesn't make sense since we only have 2 IP stacks (one is IPv4
        // and the other is IPv6)
        if (requireDistinctPorts)
          throw new IllegalArgumentException("Each listener must have a different port unless exactly one listener has " +
            s"an IPv4 address and the other IPv6 address, listeners: $listeners, port: $port")
    }
}
 ```
 
 And its still failing on the test. Lets revisit this later, I believe something is being missed.

Reply:
Sure. Thanks.